source-firestore/main.go

```diff
@@ -0,0 +1,411 @@
+			data[firestore.DocumentID] = doc.Ref.ID
+			docJson, err := json.Marshal(data)
+			docsCh <- docJson
+		}
+		close(docsCh)
+
+		if err := eg.Wait(); err != nil {
+			return err
+		}
+
+		catalog.Streams = append(catalog.Streams, airbyte.Stream{
+			Name:                    collection.ID,
+			JSONSchema:              schema,
+			SupportedSyncModes:      []airbyte.SyncMode{airbyte.SyncModeIncremental},
+			SourceDefinedCursor:     true,
+			SourceDefinedPrimaryKey: [][]string{{firestore.DocumentID}},
+		})
+	}
+
+	var encoder = airbyte.NewStdoutEncoder()
+	return encoder.Encode(airbyte.Message{
+		Type:    airbyte.MessageTypeCatalog,
+		Catalog: catalog,
+	})
+}
+
+// struct sent over channel to encoder
+type encDocument struct {
+	doc        []byte
+	streamName string
+	state      []byte
+}
+
+func doRead(args airbyte.ReadCmd) error {
+	var cfg config
+	var state State
+	var catalog airbyte.ConfiguredCatalog
+
+	if err := args.ConfigFile.Parse(&cfg); err != nil {
+		return err
+	} else if err := args.CatalogFile.Parse(&catalog); err != nil {
+		return err
+	} else if args.StateFile != "" {
+		if err := args.StateFile.Parse(&state); err != nil {
+			return err
+		}
+	}
+
+	var lastScan = time.UnixMilli(0)
+	if state.LastScan != "" {
+		var err error
+		lastScan, err = time.Parse(time.RFC3339, state.LastScan)
+		if err != nil {
+			return fmt.Errorf("parsing last scan checkpoint failed: %w", err)
+		}
+	}
+
+	var scanInterval, _ = time.ParseDuration("12h")
+	if cfg.ScanInterval != "" {
+		var err error
+		scanInterval, err = time.ParseDuration(cfg.ScanInterval)
+
+		if err != nil {
+			return fmt.Errorf("parsing scan interval failed: %w", err)
+		}
+	}
+
+	log.WithFields(log.Fields{
+		"last_scan":     lastScan,
+		"scan_interval": scanInterval,
+	}).Info("state check for scan")
+
+	ctx := context.Background()
+	sa := option.WithCredentialsJSON([]byte(cfg.CredentialsJSON))
+	app, err := firebase.NewApp(ctx, nil, sa)
+	if err != nil {
+		return err
+	}
+
+	client, err := app.Firestore(ctx)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	// The channel where documents will be sent to from different streams
+	// this is here to ensure we write documents sequentially and avoid scrambled outputs
+	var docsCh = make(chan encDocument)
+
+	var eg = new(errgroup.Group)
+
+	for _, stream := range catalog.Streams {
+		var streamName = stream.Stream.Name
+
+		var collection = client.Collection(streamName)
+		eg.Go(func() error {
+			return fullScan(ctx, lastScan, scanInterval, collection, docsCh, catalog.Tail)
+		})
+
+		if catalog.Tail {
+			eg.Go(func() error {
+				return listenForChanges(ctx, collection, docsCh)
+			})
+		}
+	}
+
+	eg.Go(func() error {
+		return encodeDocs(docsCh)
+	})
+
+	if err := eg.Wait(); err != nil {
+		return err
+	}
+
+	close(docsCh)
+
+	return nil
+}
+
+func encodeDocs(docsCh chan encDocument) error {
+	var enc = airbyte.NewStdoutEncoder()
+
+	for {
+		encDoc := <-docsCh
+
+		if len(encDoc.doc) != 0 {
```

Comment:
Question for my own enlightenment's sake:  What happens if neither of these branches applies? Is such a thing possible? If I had a better grasp on what a doc of length 0 & state of length 0 represented I might suggest adding a log for this scenario, but I'm not sure if that would make sense or not.

Reply:
I refactored to use an enum for distinguishing between different requests to encoder